frontend/src/scenes/cohorts/cohortLogic.ts

```diff
@@ -113,17 +119,82 @@ export const cohortLogic = kea<cohortLogicType<CohortLogicProps>>({
                 },
             },
         ],
+        pollTimeout: [
+            null as NodeJS.Timeout | null,
+            {
+                setPollTimeout: (_, { pollTimeout }) => pollTimeout,
+            },
+        ],
         lastSavedAt: [
             false as string | false,
             {
                 setLastSavedAt: (_, { lastSavedAt }) => lastSavedAt,
             },
         ],
-        submitted: [
-            // Indicates the form has been submitted at least once. Used to display validation errors if applicable.
-            false,
+    }),
+
+    forms: ({ actions }) => ({
+        cohort: {
+            defaults: processCohortOnSet(NEW_COHORT),
+            validator: ({ name, csv, is_static, id, groups }) => ({
+                name: !name ? 'You need to set a name' : undefined,
+                csv: id === 'new' && is_static && !csv ? 'You need to upload a CSV file' : (null as any),
+                groups: groups?.map(({ matchType, properties, action_id, event_id }) => {
+                    if (
+                        !is_static &&
+                        ((matchType === PROPERTY_MATCH_TYPE && !properties?.length) ||
+                            (matchType === ENTITY_MATCH_TYPE && !(action_id || event_id)))
+                    ) {
+                        return { id: 'This matching group is invalid' }
```

Comment:
I'm a bit confused about why we return the dict `{id:'error string'}` here instead of the error message. What you have definitely works and seems to be correct when I played with it, but I haven't wrapped my head around how kea form validators work. Mind explaining it - more so I understand than anything.

Reply:
Indeed, putting a generic error on `id` feels sort of hacky. Could there be a cleaner way? Maybe not, but if so, I'd add a disclaimer here describing the wonky behaviour 🤪

Reply:
Yeah its definitely a bit misleading since we're not necessarily erroring on this `id` field. I had to make this decision  since validator[key] must return a function returning a `DeepPartialMap` that mimics the shape of the value. 

I couldn't find a way to make this typing work without deferring the error message to `id`. Maybe `kea-forms` should also handle error strings on entire objects, but also happy to just throw in a comment for now :)

Reply:
`as any`? :)